### PR TITLE
Specify negated filters for event-level and aggregate attribution

### PR DESCRIPTION
**Pull request by linnan-github**
***
<a href="https://pr-preview.s3.amazonaws.com/WICG/attribution-reporting-api/pull/435.html" title="Last updated on Jun 2, 2022, 3:18 PM UTC (3f9b00f)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/attribution-reporting-api/435/9e434cb...3f9b00f.html" title="Last updated on Jun 2, 2022, 3:18 PM UTC (3f9b00f)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -291,10 +291,13 @@ A trigger state is a [=struct=] with the following items:
 
 A randomized source response is null or a [=set=] of [=trigger states=].
 
+<h3 dfn-type=dfn>Attribution filter value</h3>
```
Comment (johnivdel) [inline index.bs:294]: Could we drop the attribution name from these types as well?

We could maybe make this a bit more concises by doing:

<h3>Attribution filtering</h3>
A <dfn>filter value</dfn> is an [=ordered set=] of [=strings=].

A <dfn>filter map</dfn> is an [=ordered map=] whose [=map/key|keys=] are [=strings=] and whose
[=map/value|values=] are [=filter values=].

File: index.bs [hunk 2]
```diff
@@ -874,17 +873,29 @@ Given a [=request=] |request|:
 
 <h3 dfn id="does-filter-data-match">Does filter data match</h3>
 
+To <dfn>match two [=attribution filter value|filter values=]</dfn> |a| and |b| given a [=boolean=] |isNegated|:
```
Comment (johnivdel) [inline index.bs:876]: I wonder if this section would be easier to understand if we defined two different algorithms for negation?

For example:

To <dfn>match filter values</dfn> given a [=filter value=] |a| and a [=filter value=] |b|:
...

To <dfn>match filter values with negation</dfn> given a [=filter value=] |a| and a [=filter value=] |b|:

Then we could just switch/branch in the match filter data algorithm. wdyt?

File: index.bs [hunk 3]
```diff
@@ -291,10 +291,11 @@ A trigger state is a [=struct=] with the following items:
 
 A randomized source response is null or a [=set=] of [=trigger states=].
 
-<h3 dfn-type=dfn>Attribution filter map</h3>
+<h3 id="attribution-filtering">Attribution filtering</h3>
+A <dfn>filter value</dfn> is an [=ordered set=] of [=strings=].
```
Comment (apasel422) [inline index.bs:295]:
```suggestion

A <dfn>filter value</dfn> is an [=ordered set=] of [=strings=].
```

File: index.bs [hunk 4]
```diff
@@ -874,17 +871,36 @@ Given a [=request=] |request|:
...
+1. If |b| [=list/is empty=], then:
+    1. If |a| [=list/is empty=], then return true.
+    1. Otherwise, return false.
+1. Let |i| be the [=set/intersection=] of |a| and |b|.
+1. If |i| [=list/is empty=], then return false.
+1. Return true.
+
+To <dfn>match [=filter values=] with negation</dfn> given a [=filter value=] |a| and a [=filter value=] |b|:
+1. If |b| [=list/is empty=], then:
+    1. If |a| is not [=list/is empty|empty=], then return  true.
```
Comment (apasel422) [inline index.bs:884]:
```suggestion
    1. If |a| is not [=list/is empty|empty=], then return true.
```

File: index.bs [hunk 5]
```diff
@@ -874,17 +871,36 @@ Given a [=request=] |request|:
 
 <h3 dfn id="does-filter-data-match">Does filter data match</h3>
 
+To <dfn>match [=filter values=]</dfn> given a [=filter value=] |a| and a [=filter value=] |b|:
+1. If |b| [=list/is empty=], then:
```
Comment (apasel422) [inline index.bs:875]: For clarity, should we replace all these `[=list/is empty=]` with `[=set/is empty=]`? Not sure if the links work correctly, though.

File: index.bs [hunk 6]
```diff
@@ -874,17 +871,36 @@ Given a [=request=] |request|:
...
+    1. If |a| is not [=list/is empty|empty=], then return  true.
+    1. Otherwise, return false.
+1. Let |i| be the [=set/intersection=] of |a| and |b|.
+1. If |i| is not [=list/is empty|empty=], then return false.
+1. Return true.
+
 To <dfn>match an attribution source's filter data against filters</dfn> given an
-[=attribution source=] |source| and an [=attribution filter map=] |filters|:
+[=attribution source=] |source|, an [=filter map=] |filters| and a [=boolean=]
```
Comment (apasel422) [inline index.bs:891]:
```suggestion
[=attribution source=] |source|, a [=filter map=] |filters|, and a [=boolean=]
```

File: index.bs [hunk 7]
```diff
@@ -952,7 +968,11 @@ To <dfn>create [=attribution aggregatable histograms=]</dfn> given an [=attribut
 1. Let |aggregatableTrigger| be |trigger|'s [=attribution trigger/aggregatable trigger=].
 1. [=list/iterate|For each=] |triggerData| of |aggregatableTrigger|'s [=attribution aggregatable trigger/trigger data=]:
     1. If the result of running [=match an attribution source's filter data against filters=] with
-        |source| and |triggerData|'s [=attribution aggregatable trigger data/filters=] is false, [=iteration/continue=].
+        |source|, |triggerData|'s [=attribution aggregatable trigger data/filters=] and
```
Comment (apasel422) [inline index.bs:971]:
```suggestion
        |source|, |triggerData|'s [=attribution aggregatable trigger data/filters=], and
```

File: index.bs [hunk 8]
```diff
@@ -952,7 +968,11 @@ To <dfn>create [=attribution aggregatable histograms=]</dfn> given an [=attribut
...
+        [=match an attribution source's filter data against filters/isNegated=] set to false is false, [=iteration/continue=].
+    1. If the result of running [=match an attribution source's filter data against filters=] with
+        |source|, |triggerData|'s [=attribution aggregatable trigger data/negated filters=] and
```
Comment (apasel422) [inline index.bs:974]:
```suggestion
        |source|, |triggerData|'s [=attribution aggregatable trigger data/negated filters=], and
```

File: index.bs [hunk 9]
```diff
@@ -989,7 +1011,8 @@ To <dfn>trigger attribution</dfn> given an [=attribution trigger=] |trigger| run
     null or an [=set/is empty|empty=] [=set=].
 1. If the result of running
     [=match an attribution source's filter data against filters=] with
-    |sourceToAttribute| and |trigger|'s [=attribution trigger/filters=] is false, return.
+    |sourceToAttribute|, |trigger|'s [=attribution trigger/filters=] and
```
Comment (apasel422) [inline index.bs:1014]:
```suggestion
    |sourceToAttribute|, |trigger|'s [=attribution trigger/filters=], and
```

File: index.bs [hunk 10]
```diff
@@ -1002,8 +1025,13 @@ To <dfn>trigger attribution</dfn> given an [=attribution trigger=] |trigger| run
 1. [=map/iterate|For each=] [=event-level trigger configuration=] |config| of |trigger|'s
     [=attribution trigger/event-level trigger configurations=]:
     1. If the result of running
-        [=match an attribution source's filter data against filters=] with |sourceToAttribute| and
-        |config|'s [=event-level trigger configuration/filters=] is false, [=iteration/continue=].
+        [=match an attribution source's filter data against filters=] with |sourceToAttribute|,
+        |config|'s [=event-level trigger configuration/filters=] and
```
Comment (apasel422) [inline index.bs:1029]:
```suggestion
        |config|'s [=event-level trigger configuration/filters=], and
```

File: index.bs [hunk 11]
```diff
@@ -1002,8 +1025,13 @@ To <dfn>trigger attribution</dfn> given an [=attribution trigger=] |trigger| run
...
+        [=match an attribution source's filter data against filters/isNegated=] set to false is false, [=iteration/continue=].
+    1. If the result of running
+        [=match an attribution source's filter data against filters=] with |sourceToAttribute|,
+        |config|'s [=event-level trigger configuration/negated filters=] and
```
Comment (apasel422) [inline index.bs:1033]:
```suggestion
        |config|'s [=event-level trigger configuration/negated filters=], and
```

**Review by apasel422 — CHANGES_REQUESTED** [review]

File: index.bs (hunk 5 above)
Comment (linnan-github) [reply index.bs]: Yes, [=set/is empty=] is more clear and the link works. Just updated. Thanks.

**Review by apasel422 — APPROVED** [review]

**Review by johnivdel — APPROVED** [review]
lgtm with suggestion

File: index.bs [hunk 12]
```diff
@@ -874,17 +872,36 @@ Given a [=request=] |request|:
...
 
 1. Let |sourceData| be |source|'s [=attribution source/filter data=].
 1. [=map/iterate|For each=] |key| → |filterValues| of |filters|:
     1. If |sourceData|[|key|] does not [=map/exist=], [=iteration/continue=].
     1. Let |sourceValues| be |sourceData|[|key|].
-    1. If |sourceValues| [=list/is empty=] and |filterValues| is [=list/is empty=],
-        [=iteration/continue=].
-    1. Let |i| be the [=set/intersection=] of |sourceValues| and |filterValues|.
-    1. If |i| [=list/is empty=], return false.
+    1. If |isNegated| is false, then:
```
[the remark this answers is not in the document]
Comment (linnan-github) [reply index.bs:898]: Thanks John. Updated with switch, thanks.

**Review by johnivdel — APPROVED** [review]
slgtm

**Pull request merged**